subsys/bluetooth/services/ias/ias_client.c

```diff
@@ -0,0 +1,241 @@
+
+	client_cleanup(client_by_conn(conn));
+	ias_client_cb->discover(conn, err);
+}
+
+int bt_ias_client_alert_high(struct bt_conn *conn)
+{
+	int err;
+	const uint8_t lvl = BT_IAS_ALERT_LVL_HIGH_ALERT;
+
+	CHECKIF(conn == NULL) {
+		return -ENOTCONN;
+	}
+
+	if (client_by_conn(conn)->write_handle == 0) {
+		return BT_ATT_ERR_INVALID_HANDLE;
+	}
+
+	err = bt_gatt_write_without_response(conn,
+					client_by_conn(conn)->write_handle,
+					&lvl, sizeof(lvl), false);
+	if (err < 0) {
+		BT_ERR("IAS client high alert level write failed: %d", err);
+
+		return err;
+	}
+
+	return err;
+}
+
+int bt_ias_client_alert_mild(struct bt_conn *conn)
+{
+	int err;
+	const uint8_t lvl = BT_IAS_ALERT_LVL_MILD_ALERT;
+
+	CHECKIF(conn == NULL) {
+		return -ENOTCONN;
+	}
+
+	if (client_by_conn(conn)->write_handle == 0) {
+		return BT_ATT_ERR_INVALID_HANDLE;
+	}
+
+	err = bt_gatt_write_without_response(conn,
+					client_by_conn(conn)->write_handle,
+					&lvl, sizeof(lvl), false);
+	if (err < 0) {
+		BT_ERR("IAS client mild alert level write failed: %d", err);
+
+		return err;
+	}
+
+	return err;
+}
+
+int bt_ias_client_alert_stop(struct bt_conn *conn)
+{
+	int err;
+	const uint8_t lvl = BT_IAS_ALERT_LVL_NO_ALERT;
+
+	CHECKIF(conn == NULL) {
+		return -ENOTCONN;
+	}
+
+	if (client_by_conn(conn)->write_handle == 0) {
+		return BT_ATT_ERR_INVALID_HANDLE;
+	}
+
+	err = bt_gatt_write_without_response(conn,
+					client_by_conn(conn)->write_handle,
+					&lvl, sizeof(lvl), false);
+	if (err < 0) {
+		BT_ERR("IAS client stop alert write failed: %d", err);
+
+		return err;
+	}
+
+	return err;
+}
+
+static uint8_t bt_ias_alert_lvl_disc_cb(struct bt_conn *conn,
+				const struct bt_gatt_attr *attr,
+				struct bt_gatt_discover_params *discover)
+{
+	const struct bt_gatt_chrc *chrc = (struct bt_gatt_chrc *)attr->user_data;
+
+	atomic_clear_bit(client_by_conn(conn)->flags, IAS_DISCOVER_IN_PROGRSS);
+
+	if (attr == NULL) {
+		ias_client_cb->discover(conn, -BT_ATT_ERR_ATTRIBUTE_NOT_FOUND);
+
+		return BT_ATT_ERR_ATTRIBUTE_NOT_FOUND;
+	}
+
+	client_by_conn(conn)->write_handle = chrc->value_handle;
+
+	return BT_GATT_ITER_STOP;
+}
+
+static uint8_t bt_ias_prim_disc_cb(struct bt_conn *conn,
+				const struct bt_gatt_attr *attr,
+				struct bt_gatt_discover_params *discover)
+{
+	int err;
+	const struct bt_gatt_service_val *data;
+	struct bt_ias_client *client = client_by_conn(conn);
+
+	if (!attr) {
+		BT_DBG("Discover complete\n");
+
+		return BT_GATT_ITER_STOP;
+	}
+
+	data = attr->user_data;
+
+	client->discover.uuid = alert_lvl_uuid;
+	client->discover.start_handle = attr->handle + 1;
+	client->discover.end_handle = data->end_handle;
+	client->discover.type = BT_GATT_DISCOVER_CHARACTERISTIC;
+	client->discover.func = bt_ias_alert_lvl_disc_cb;
+
+	err = bt_gatt_discover(conn, &client->discover);
+	if (err) {
+		BT_ERR("Characteristic discover failed\n");
+		discover_failed(conn, err);
+
+		return BT_GATT_ITER_STOP;
+	}
+
+	ias_client_cb->discover(conn, err);
```

Comment:
Are you sure you want to call `discover` callback here and not wait for `bt_ias_alert_lvl_disc_cb` ?

Reply:
I think you should